rust/plugin/src/diff/differ.rs

```diff
@@ -1,126 +1,317 @@
+    }
+
+	/// Computes the diff between the two sets of heads.
+    #[instrument(skip_all, level = tracing::Level::DEBUG)]
+    pub fn get_diff(&self) -> ProjectDiff {
+        tracing::debug!(
+            "branch {:?}, getting changes between {} and {}",
+            self.branch_state.name,
+            self.prev_heads.to_short_form(),
+            self.curr_heads.to_short_form()
+        );
+
+        if self.prev_heads == self.curr_heads {
+            tracing::debug!("no changes");
+            return ProjectDiff::default();
+        }
+
+        let mut diffs: Vec<Diff> = vec![];
+        // Get old and new content
+        let new_file_contents = self.project.get_changed_file_content_between(
+            None,
+            self.branch_state.doc_handle.document_id().clone(),
+            self.prev_heads.clone(),
+            self.curr_heads.clone(),
+            false,
+        );
+        let changed_files_set: HashSet<String> = new_file_contents
+            .iter()
+            .map(|event| match event {
+                FileSystemEvent::FileCreated(path, _) => path.to_string_lossy().to_string(),
+                FileSystemEvent::FileModified(path, _) => path.to_string_lossy().to_string(),
+                FileSystemEvent::FileDeleted(path) => path.to_string_lossy().to_string(),
+            })
+            .collect::<HashSet<String>>();
+        let old_file_contents = self.project.get_files_on_branch_at(
+            self.branch_state,
+            Some(&self.prev_heads),
+            Some(&changed_files_set),
+        );
+
+        for event in &new_file_contents {
+            let (path, new_file_content, change_type) = match event {
+                FileSystemEvent::FileCreated(path, content) => (path, content, ChangeType::Added),
+                FileSystemEvent::FileModified(path, content) => {
+                    (path, content, ChangeType::Modified)
+                }
+                FileSystemEvent::FileDeleted(path) => {
+                    (path, &FileContent::Deleted, ChangeType::Removed)
+                }
+            };
+            let path = path.to_string_lossy().to_string();
+            let old_file_content = old_file_contents
+                .get(&path)
+                .unwrap_or(&FileContent::Deleted);
+            let old_content_type = old_file_content.get_variant_type();
+            let new_content_type = new_file_content.get_variant_type();
+            let old_content = match old_content_type {
+                VariantType::NIL => None,
+                _ => Some(old_file_content),
+            };
+            let new_content = match new_content_type {
+                VariantType::NIL => None,
+                _ => Some(new_file_content),
+            };
+
+            if old_content_type == VariantType::OBJECT || new_content_type == VariantType::OBJECT {
+                // This is a scene file, so use a scene diff
+                diffs.push(Diff::Scene(self.get_scene_diff(&path)));
```

Comment:
given that we've already loaded both the new content and old content at this point, let's just pass it into `get_scene_diff()`. Hydrating scenes at heads is expensive, and will be more expensive when we upgrade automerge